src/ethereum/berlin/eth_types.py

```diff
@@ -0,0 +1,204 @@
+"""
+Ethereum Types
+^^^^^^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+Types re-used throughout the specification, which are specific to Ethereum.
+"""
+
+from dataclasses import dataclass
+from typing import List, Tuple, Union
+
+from .. import rlp
+from ..base_types import (
+    U256,
+    Bytes,
+    Bytes0,
+    Bytes8,
+    Bytes20,
+    Bytes32,
+    Bytes256,
+    Uint,
+    Uint64,
+    slotted_freezable,
+)
+from ..crypto.hash import Hash32, keccak256
+from ..exceptions import InvalidBlock
+from ..utils.ensure import ensure
+
+Address = Bytes20
+Root = Hash32
+
+Bloom = Bytes256
+
+TX_BASE_COST = 21000
+TX_DATA_COST_PER_NON_ZERO = 16
+TX_DATA_COST_PER_ZERO = 4
+TX_CREATE_COST = 32000
+TX_ACCESS_LIST_ADDRESS_COST = 2400
+TX_ACCESS_LIST_STORAGE_KEY_COST = 1900
+
+
+@slotted_freezable
+@dataclass
+class LegacyTransaction:
+    """
+    Atomic operation performed on the block chain.
+    """
+
+    nonce: U256
+    gas_price: U256
+    gas: U256
+    to: Union[Bytes0, Address]
+    value: U256
+    data: Bytes
+    v: U256
+    r: U256
+    s: U256
+
+
+@slotted_freezable
+@dataclass
+class Transaction2930:
+    """F"""
+
+    chain_id: Uint64
+    nonce: U256
+    gas_price: U256
+    gas: U256
+    to: Union[Bytes0, Address]
+    value: U256
+    data: Bytes
+    access_list: List[Tuple[Address, List[Bytes32]]]
+    v: U256
+    r: U256
+    s: U256
+
+
+Transaction = Union[LegacyTransaction, Transaction2930]
+
+
+def encode_transaction(tx: Transaction) -> Union[LegacyTransaction, Bytes]:
```

Comment:
Yes. EIP-2718 specifies that a transaction is either:

1. A `LegacyTransaction` RLP object, i.e. an RLP list of stuff.
2. A `Bytes` consisting of a version byte followed by version defined opaque bytes, i.e. an RLP encode bytestring.

The subtle point here is that `legacy_transaction` and `rlp.encode(legacy_transaction)` are different. This doesn't matter in `transactions_trie`, because `legacy_transaction` and `rlp.encode(legacy_transaction)` have the same encoding in tries. It does matter when dealing with block RLP (found in `debug_getBlockRlp` and the testsuite) though.

Reply:
> I'm still rather confused because both branches seem to return approximately `rlp.encode(...)`, which is `Bytes`?

That is a bug. I will fix it and add a test to check that `encode_transaction()` and `decode_transaction()` are opposites.

Reply:
> That is a bug.

I'm not totally useless :tada: